recipe/bld.bat

```diff
@@ -1,50 +1,195 @@
+    set TORCH_CUDA_ARCH_LIST=3.5;5.0+PTX
+    if "%cuda_compiler_version%" == "11.8" (
+        set TORCH_CUDA_ARCH_LIST=3.5;5.0;6.0;6.1;7.0;7.5;8.0;8.6;8.9+PTX
+        set CUDA_TOOLKIT_ROOT_DIR=%CUDA_HOME%
+    ) else if "%cuda_compiler_version%" == "12.0" (
+        set TORCH_CUDA_ARCH_LIST=5.0;6.0;6.1;7.0;7.5;8.0;8.6;8.9;9.0+PTX
+        set CUDA_TOOLKIT_ROOT_DIR=%PREFIX%
+    ) else (
+        echo "unsupported cuda version. edit build_pytorch.bat"
+        exit /b 1
+    )
 
-set MAGMA_HOME=%LIBRARY_PREFIX%
+    set TORCH_NVCC_FLAGS=-Xfatbin -compress-all
 
-set "PATH=%CUDA_BIN_PATH%;%PATH%"
+    REM set USE_SYSTEM_NCCL=1
+    set USE_STATIC_NCCL=0
+    set USE_STATIC_CUDNN=0
+    set MAGMA_HOME=%PREFIX%
 
-set CUDNN_INCLUDE_DIR=%LIBRARY_PREFIX%\include
+    REM NCCL is not available on windows
+    set USE_NCCL=0
+
+    set MAGMA_HOME=%LIBRARY_PREFIX%
+
+    set "PATH=%CUDA_BIN_PATH%;%PATH%"
+
+    set CUDNN_INCLUDE_DIR=%LIBRARY_PREFIX%\include
+
+) else (
+    set USE_CUDA=0
+    
+    @REM MKLDNN is an Apache-2.0 licensed library for DNNs and is used
+    @REM for CPU builds. Not to be confused with MKL.
+    set "USE_MKLDNN=1"
+)
+
+set DISTUTILS_USE_SDK=1
 
-:cuda_end
+set CMAKE_INCLUDE_PATH=%LIBRARY_PREFIX%\include
+set LIB=%LIBRARY_PREFIX%\lib;%LIB%
 
+@REM CMake configuration
 set CMAKE_GENERATOR=Ninja
+set "CMAKE_GENERATOR_TOOLSET="
 set "CMAKE_GENERATOR_PLATFORM="
 set "CMAKE_PREFIX_PATH=%LIBRARY_PREFIX%"
+set "CMAKE_INCLUDE_PATH=%LIBRARY_INC%"
+set "CMAKE_LIBRARY_PATH=%LIBRARY_LIB%"
+set "CMAKE_BUILD_TYPE=Release"
+
+set "INSTALL_TEST=0"
+set "BUILD_TEST=0"
+
 set "libuv_ROOT=%LIBRARY_PREFIX%"
-set "USE_SYSTEM_SLEEF=OFF"
-set "BUILD_CUSTOM_PROTOBUF=OFF"
+set "USE_SYSTEM_SLEEF=ON"
+
+@REM uncomment to debug cmake build
+@REM set "CMAKE_VERBOSE_MAKEFILE=1"
+
+@REM TODO(baszalmstra): There are link errors because of conflicting symbols with caffe2_protos.lib
+set "BUILD_CUSTOM_PROTOBUF=ON"
+
+@REM TODO(baszalmstra): There are linker errors because of mixing Intel OpenMP (iomp) and Microsoft OpenMP (vcomp)
+set "USE_OPENMP=0"
+
+@REM The activation script for cuda-nvcc doesnt add the CUDA_CFLAGS on windows. 
+@REM Therefor we do this manually here. See:
+@REM https://github.com/conda-forge/cuda-nvcc-feedstock/issues/47
+echo "CUDA_CFLAGS=%CUDA_CFLAGS%"
+set "CUDA_CFLAGS=-I%PREFIX%/Library/include -I%BUILD_PREFIX%/Library/include"
+set "CFLAGS=%CFLAGS% %CUDA_CFLAGS%"
+set "CPPFLAGS=%CPPFLAGS% %CUDA_CFLAGS%"
+set "CXXFLAGS=%CXXFLAGS% %CUDA_CFLAGS%"
+echo "CUDA_CFLAGS=%CUDA_CFLAGS%"
+echo "CXXFLAGS=%CXXFLAGS%"
+
+@REM Configure sccache
+set "CMAKE_C_COMPILER_LAUNCHER=sccache"
+set "CMAKE_CXX_COMPILER_LAUNCHER=sccache"
+set "CMAKE_CUDA_COMPILER_LAUNCHER=sccache"
+
+sccache --stop-server
+sccache --start-server
+sccache --zero-stats
 
-%PYTHON% -m pip install . --no-deps -vv
+@REM Clear the build from any remaining artifacts. We use sccache to avoid recompiling similar code.
+cmake --build build --target clean
```

Comment:
Why are you cleaning here? Unix builds do not.

Reply:
Good question. The reason is that we get into a weird state on Windows where apparently the order of the variant builds for pytorch influences the compilation process. I noticed that this order is non-deterministic. We would run into a weird issue where if pytorch for python 3.11 was built before 3.12 you will get linker errors for the 3.12 build. Whereas if 3.12 would be built before 3.11 all would be good.

I have been trying to figure out why this happens and why this specifically only happens for Python 3.11 and 3.12 but I couldnt figure it out. As a last resort, I opted to just remove all artifacts and rebuild them. Using sccache to ensure we aren't rebuilding everything from scratch.

This is a little bit of a temporary workaround. As far as I understand with [pytorch 2.4.1 split builds will be supported out of the box](https://github.com/pytorch/pytorch/pull/126328) and all of the megabuild stuff will become a lot simpler.

If you happen to have a better idea though, I welcome it with open arms! 😄

Reply:
Do you have a log with the linker errors?

> As far as I understand with https://github.com/pytorch/pytorch/pull/126328 and all of the megabuild stuff will become a lot simpler.

It probably won't be ready for use until 2.5.0. Interesting note: pytorch developers wanted to replicate the libtorch split in conda-forge in wheels.

Reply:
Yes the last time this failed in CI was here: https://github.com/conda-forge/pytorch-cpu-feedstock/actions/runs/10077183574/job/27859126386?pr=231

And interesting indeed! That's great. 👍

Reply:
Can you print build/CMakeCache.txt ?